swarm/hash/types.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2018 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swarmhash
+
+import (
+	"fmt"
+	"hash"
+	"sync"
+)
+
+const (
+	DefaultHasherCount = 32 // default number of workers to instantiate in a pool upon creation
+)
+
+var (
+	defaultHash string
+	pools       map[string]*Hasher
+	initFunc    sync.Once
+)
+
+// the basic swarm hash interface
+// an instance represents an individual worker in a pool
+type Hash interface {
+	hash.Hash
+	ResetWithLength([]byte)
+}
+
+// function for creating a new individual worker
+type HashFunc func() Hash
+
+// Hasher maintains a pool of workers
+// It is the entrypoint for invoking hashing jobs
+// And to query metadata about the type of worker in its pool
+type Hasher struct {
+	typ      string
+	hashFunc func() Hash
+	pool     sync.Pool
+	size     int
+}
+
+func init() {
+	pools = make(map[string]*Hasher)
+}
+
+// Init sets the default hash to be used when calling GetHash
+// This must be executed before any hashing can commence
+// If no hash functions is registered with the identifier, the function will panic
+func Init(typ string) {
```

Comment:
this is quite ugly to initialise from a global default const
let make a contructor then, and pass argument in the code.